### DISCUSSION TITLE
What does upsert actually return?

### DISCUSSION THREAD

**Discussion by xorock** [laravel/framework#34979] [Q&A] [answered]
Hi, I have 5 db records (mysql) with given PK (compound, on 2 columns). And I have 5 records that require update / insert. If I run upsert on all records it returns (int) 10 affected rows. If I delete one db record (4) and upsert 5 records it returns (int) 9. Shouldn't it be 5 in both cases?

**Comment by paras-malhotra** [comment, marked as answer]
@xorock the `upsert` function returns what the database engine returns for upserts. This can be specific to the DB engine you're using because upserts aren't really SQL standards.

So, for instance, if you're using mysql, here's how it calculates the number of affected rows (direct quote from [docs](https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html)):

> With ON DUPLICATE KEY UPDATE, the affected-rows value per row is 1 if the row is inserted as a new row, 2 if an existing row is updated, and 0 if an existing row is set to its current values.

**Reply by anstapol** [reply]
So basically the returned number should never be lower than the length of the data array for models with timestamps enabled right?

Because per mass update the values for timestamps would be different, meaning it'll update anyways right?
(direct quote from [docs](https://laravel.com/docs/10.x/eloquent#upserts)):

>   The upsert method will automatically set the created_at and updated_at timestamps if timestamps are enabled on the model.

**Comment by johnnybenson** [comment]
This information should really be included in the documentation https://laravel.com/docs/10.x/eloquent#upserts
